### ISSUE TITLE
[Bounty $1500] Conv3d T dim padding

### ISSUE THREAD

**Issue by astancovTT** [tenstorrent/tt-metal#25936]
## 📝 Background

Currently, `ttnn.experimental.conv3d` in tt-metal does not support padding along the T (temporal) dimension, which is a standard parameter in 3D convolution operations. This limitation is enforced by an assertion that requires `padding[0] == 0`, effectively blocking temporal padding. Supporting padding along the T dimension is crucial for applications like video analysis, volumetric medical imaging, and 3D object detection—where proper temporal context is necessary for correctness and generality.

## 🎯 What Success Looks Like

- Remove the restriction that forces `padding[0] == 0` in the device code.
- Update kernel implementation to correctly handle non-zero T-dimension (temporal) padding.
- Output dimension calculations account for T-dimension padding.
- API accepts the same padding parameter format as PyTorch's `torch.nn.Conv3d`, including T-dimension (D in PyTorch) padding.
- Support both "zeros" and "replicate" padding modes for the T dimension.
- Comprehensive tests added to `tests/ttnn/nightly/unit_tests/operations/conv/test_conv3d.py` validating numerical correctness against PyTorch Conv3d for several non-trivial temporal padding cases.

## 💡 Problem to Solve

The current tt-metal Conv3d device implementation blocks T-dimension padding by assertion, even though the underlying config and kernel structure could support it. PyTorch’s Conv3d supports arbitrary padding, including the depth/temporal dimension. This bounty requires updating the device operation and kernel logic to permit and correctly process temporal padding—making the implementation correct and much more broadly applicable.

## 🧭 Guidance & Starting Points

- See the assertion in `ttnn/cpp/ttnn/operations/experimental/conv3d/device/conv3d_device_operation.cpp`:
  ```cpp
  // assert padding on T is zero
  TT_FATAL(
      config.padding[0] == 0,
      "Padding must be (0,x,x). got ({}, {}, {})",
      config.padding[0],
      config.padding[1],
      config.padding[2]);
  ```
- Kernel logic for vol2col in `ttnn/cpp/ttnn/operations/experimental/conv3d/device/reader_vol2col.cpp` already has a framework for padding handling, but is restricted - ensure T-dimension logic is complete.
- Output shape calculation must include T-dimension padding.
- Test coverage in `tests/ttnn/nightly/unit_tests/operations/conv/test_conv3d.py`—add cases such as:
  - `padding=(1,1,1)`, `padding=(2,0,1)`, `padding=(3,2,1)`
  - Both "zeros" and "replicate" T-dimension padding
  - Compare output to PyTorch Conv3d

## 🔎 Possible Approaches

- Remove the assertion enforcing zero T padding.
- Read and update the kernel logic for proper handling of temporal padding. Follow the pattern for H/W padding.
- Refactor or extend output dimension calculation to match PyTorch’s convention.
- Use PyTorch Conv3d as reference for expected behavior and output.
- Write/extend unit tests to confirm correctness for typical and edge-case padding values.

---
## 📚 Resources

- [PyTorch Conv3d Documentation](https://pytorch.org/docs/stable/generated/torch.nn.Conv3d.html)
- Existing Conv2d implementation for reference patterns in padding logic
- Relevant tt-metal files:
  - [`conv3d_device_operation.cpp`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/experimental/conv3d/device/conv3d_device_operation.cpp)
  - [`reader_vol2col.cpp`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/experimental/conv3d/device/reader_vol2col.cpp)
  - [`test_conv3d.py`](https://github.com/tenstorrent/tt-metal/blob/main/tests/ttnn/nightly/unit_tests/operations/conv/test_conv3d.py)

**Comment by RafLit** [comment]
Hi, I’d be happy to be assigned to this issue.

**Comment by Yashashwini2003** [comment]
Hi @marty1885 I'm interested to take this up

**Comment by marty1885** [comment]
Assigning to @RafLit - do you have access to our hardware? We can provide free cloud access if you need one.

Either case @RafLit, it's yours now, thank you for the interest! Please let us know within 2 weeks your progress. Please refer to our [Bounty Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html). Good luck with the project! And let us know if there's any bug/issues on our side that is stopping you find making the bounty happen. And note as Tenstorrent is an US based company. If you are an non-US resided participants, unless the country of your residence have a tax treaty with the US. A 30% withholding will be deduced upon payout for tax withholding. Please let me know if you have any questions and if I can help.

**Comment by RafLit** [comment]
@marty1885 Thank you, will update soon!
I actually have the hardware and will use it, but I appreciate the possibility to use cloud access

**Comment by RafLit** [comment]
Update:
-I familiarized myself with the metalium api by following programming examples and trying some of my own
-I familiarized myself with the code of the conv3d operation and the mentioned reader kernel - for now it looks to me like it treats the temporal dimension the same as both H and W - the padding function goes through it as well.
-I made a draft PR [#31720](https://github.com/tenstorrent/tt-metal/pull/31720) where i removed the assertion for the T-dim padding being equal to zero and added some tests with padding>0.
-I ran all those tests and all of them seem to pass, but I want to make some additional verification.

**Comment by marty1885** [comment]
Thank your for the update! Really appreciate it! LMK if there's anything I can help with.

**Comment by RafLit** [comment]
I did some additional verification and still all tests seem to pass :)
I will cleanup the tests and make the pr review-ready now

**Comment by marty1885** [comment]
How is this going?

**Comment by mert-kurttutan** [comment]
What is the progress on the issue? This is also important for other projects makign use of Conv3d. This is because of many of them are using kernel size 3 with the padding=`same` (meaning padding 1 on each side). So, one has to manually pad before the conv3d operation.
It would be great if we could have some update.

**Comment by marty1885** [comment]
Ping @RafLit ~

**Comment by RafLit** [comment]
@marty1885  @mert-kurttutan  Hi, really sorry for the late response.
I updated the pull request here: https://github.com/tenstorrent/tt-metal/pull/31720.

I updated the sweep_shape tests by adding some padding cases, but those will take more time to verify because of other parametrized variables. If this is a problem I can move this to a separate sweep_padding test, or is it ok as it is?
Otherwise i think it is ready for review.

**Comment by jberkowitzTT** [comment]
Hi @RafLit - please email me at jberkowitz@tenstorrent.com with this link or the bounty issue title to kick off the payment process

Thank you!
